go-runtime/compile/build.go

```diff
@@ -404,6 +480,237 @@ func SyncGeneratedStubReferences(ctx context.Context, projectRootDir string, stu
+		imports:     imports(result.Module, false),
+	}
+	return builder.build(goModVersion, ftlVersion, projectName, sharedModulesPaths, replacements)
+}
+
+func (b *mainModuleContextBuilder) build(goModVersion, ftlVersion, projectName string,
+	sharedModulesPaths []string, replacements []*modfile.Replace) (mainModuleContext, error) {
+	ctx := mainModuleContext{
+		GoVersion:          goModVersion,
+		FTLVersion:         ftlVersion,
+		Name:               b.mainModule.Name,
+		SharedModulesPaths: sharedModulesPaths,
+		Replacements:       replacements,
+		Verbs:              make([]goVerb, 0, len(b.mainModule.Decls)),
+		MainCtx: mainFileContext{
+			ProjectName:   projectName,
+			SumTypes:      []goSumType{},
+			ExternalTypes: []goExternalType{},
+		},
+		TypesCtx: typesFileContext{
+			SumTypes:      []goSumType{},
+			ExternalTypes: []goExternalType{},
+		},
+	}
+
+	err := schema.Visit(b.mainModule, func(node schema.Node, next func() error) error {
+		if m, ok := node.(*schema.Module); ok && m.Name != b.currentModule {
+			b.currentModule = m.Name
+			return next()
+		}
+
+		maybeGoType, isLocal, err := b.getGoType(node)
+		if err != nil {
+			return err
+		}
+		gotype, ok := maybeGoType.Get()
+		if !ok {
+			return next()
+		}
+		nt := gotype.getNativeType()
+		b.imports = addImports(b.imports, nt)
+		// update with the resolved directory name, if necessary
+		if dirName, ok := b.imports[nt.importPath]; ok {
+			if existingDirName, ok := nt.directoryName.Get(); !ok || existingDirName != dirName {
+				gotype.setDirectoryName(dirName)
+			}
+		}
+
+		switch n := gotype.(type) {
+		case goVerb:
+			ctx.Verbs = append(ctx.Verbs, n)
+		case goSumType:
+			if isLocal {
+				ctx.TypesCtx.SumTypes = append(ctx.TypesCtx.SumTypes, n)
+			}
+			ctx.MainCtx.SumTypes = append(ctx.MainCtx.SumTypes, n)
+		case goExternalType:
+			if isLocal {
+				ctx.TypesCtx.ExternalTypes = append(ctx.TypesCtx.ExternalTypes, n)
+			}
+			ctx.MainCtx.ExternalTypes = append(ctx.MainCtx.ExternalTypes, n)
+		}
+		return next()
+	})
+	if err != nil {
+		return mainModuleContext{}, fmt.Errorf("failed to build main module context: %w", err)
+	}
+
+	slices.SortFunc(ctx.MainCtx.SumTypes, func(a, b goSumType) int {
+		return strings.Compare(a.TypeName(), b.TypeName())
+	})
+	slices.SortFunc(ctx.TypesCtx.SumTypes, func(a, b goSumType) int {
+		return strings.Compare(a.TypeName(), b.TypeName())
+	})
+
+	return ctx.withImports(), nil
+}
+
+func (b *mainModuleContextBuilder) getGoType(node schema.Node) (gotype optional.Option[goType], isLocal bool, err error) {
+	switch n := node.(type) {
+	case *schema.Ref:
+		if n.Module != "" && n.Module != b.mainModule.Name {
+			return optional.None[goType](), b.visitingMainModule(), nil
+		}
+
+		resolved, ok := b.sch.Resolve(n).Get()
+		if !ok {
+			return optional.None[goType](), b.visitingMainModule(), nil
+		}
+
+		// update the current module for the visit to the resolved node
+		currentModule := b.currentModule
+		b.currentModule = n.Module
```

Comment:
Would it make sense to refactor this slightly so that the caller of `getGoType` doesn't need to maintain the `currentModule` state, and instead, `getGoType` does that itself?

Reply:
you mean passing the current module name through `getGoType` rather than keeping it in the `mainModuleContextBuilder`? i was going back and forth on which approach would read less confusing, but i think you're right! i'll update